src/webui/api/torrentscontroller.cpp

```diff
@@ -679,6 +706,58 @@ void TorrentsController::removeTrackersAction()
         torrent->forceReannounce();
 }
 
+void TorrentsController::addPeersAction()
+{
+    checkParams({"hash", "peers"});
+
+    const QString hash = params()["hash"];
+    const QStringList peers = params()["peers"].split('|');
+
+    BitTorrent::TorrentHandle *const torrent = BitTorrent::Session::instance()->findTorrent(hash);
+    if (!torrent)
+        throw APIError(APIErrorType::NotFound);
+
+    QList<BitTorrent::PeerAddress> peerList;
+    for (const QString &peer : peers) {
+        const BitTorrent::PeerAddress addr = parsePeer(peer);
+        if (addr.ip.isNull())
+            throw APIError(APIErrorType::BadParams);
+        peerList.append(addr);
+    }
+
+    int peersAdded = 0;
+    for (const BitTorrent::PeerAddress &peer : peerList) {
+        if (torrent->connectPeer(peer))
+            ++peersAdded;
+    }
+    if (peersAdded == 0)
+        throw APIError(APIErrorType::Conflict, tr("No peers were added"));
+}
+
+void TorrentsController::banPeersAction()
```

Comment:
Why is it in TorrentsController? Why is meaningless/confusing "hash" parameter needed while qBittorrent allows you to ban peers globally only?

Reply:
IMO, TransferController is better place.